### ISSUE TITLE
Declared instance variable annotation overridden by inferred type

### ISSUE THREAD

**Issue by XuaTheGrate** [microsoft/pyright#3209]
**Describe the bug**
When defining an instance variable with a type annotation, and setting the variable to an incompatible value, the annotation is overridden to match that of the incompatible value.

**To Reproduce**
```python
class Dummy:
    def __init__(self):
        self.k: int = "1"  # type: ignore
        reveal_type(self.k)  # Type of "self.k" is "Literal['1']"
```

**Expected behavior**
For the revealed type of ``self.k`` to be ``int``.

**VS Code extension or command-line**
Pylance v2022.3.2

**Additional context**
Unbound variables aren't affected by this bug, and they behave as expected:
```python
k: int = "1"  # type: ignore
reveal_type(k)  # Type of "k" is "int"
```

**Comment by erictraut** [comment]
This is intended behavior. 

First, this is a type violation, so you should not `# type: ignore` it. 

Second, within a local execution context, type narrowing is applied for assignments. In the case of an assignment that violates a type violation, we still apply the narrowing for the assignment. This is useful in cases where users are not interested in type checking but are instead more interested in completion suggestions and other language server features.

It looks like you're interested in type checking because you've added an explicit type annotation for the instance variable, but you've violated the type and then told the type checker to suppress the error generated by the type violation.

**Comment by XuaTheGrate** [comment]
The type violation is irrelevant to the issue, as that's not what causes the inconsistency. The behaviour I had shown was how it used to work (I'm ~90% sure anyway). Why is such an inconsistency "as designed"?

**Comment by XuaTheGrate** [comment]
To append to this: Mypy outputs the correct behaviour:
```
$ mypy sample.py
sample.py:2: note: Revealed type is "builtins.int"
sample.py:7: note: Revealed type is "builtins.int"

$ pyright sample.py
./sample.py
  ./sample.py:2:13 - information: Type of "k" is "int"
  ./sample.py:7:21 - information: Type of "self.k" is "Literal['1']"
```

**Comment by erictraut** [comment]
We haven't changed this behavior in a long time — probably a year ago. As I explained, this behavior is intended to work better in cases where people don't care about type checking and want to receive the most accurate completion suggestions. Pyright is used as both a type checker and a language server, so it needs to balance the needs of both. Mypy is a type checker only.

If you fix the type violation in your code, there won't be an issue.
